src/abstractdataframe/abstractdataframe.jl

```diff
@@ -1450,6 +1450,150 @@ julia> unique!(df)  # modifies df
+    duplicates are allowed. They are not repeated if `allcols` is `false`
+    only unique combinations are produced then, but if `allcols` is `true`
+    the duplicates are included.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(x=1:2, y='a':'b', z=["x", "y"])
+2×3 DataFrame
+ Row │ x      y     z
+     │ Int64  Char  String
+─────┼─────────────────────
+   1 │     1  a     x
+   2 │     2  b     y
+
+julia> completecombinations(df, [:x, :y])
+4×2 DataFrame
+ Row │ x      y
+     │ Int64  Char
+─────┼─────────────
+   1 │     1  a
+   2 │     2  a
+   3 │     1  b
+   4 │     2  b
+
+julia> completecombinations(df, [:y, :z], allcols=true)
+4×3 DataFrame
+ Row │ y     z       x
+     │ Char  String  Int64?
+─────┼───────────────────────
+   1 │ a     x             1
+   2 │ b     x       missing
+   3 │ a     y       missing
+   4 │ b     y             2
+```
+"""
+function completecombinations(df::AbstractDataFrame, indexcols; allcols::Bool=false,
+                allowduplicates::Bool=false, fill=missing)
+    colind = index(df)[indexcols]
+
+    if length(colind) == 0
+        throw(ArgumentError("At least one column to complete combinations required"))
+    end
+
+    has_duplicates = row_group_slots(ntuple(i -> df[!, colind[i]], length(colind)),
+                                     Val(false), nothing, false, nothing)[1] != nrow(df)
+    if has_duplicates && !allowduplicates
+        throw(ArgumentError("duplicate rows in input"))
+    end
+
+    # Create a vector of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        # All levels are retained, missing is added only if present
+        if any(ismissing, df[!, col])
+            tempcol = vcat(levels(df[!, col]), missing)
+        else
+            tempcol = levels(df[!, col])
+        end
+        push!(uniqueVals, tempcol)
+    end
+
+    # make sure we do not overflow in the target data frame size
+    target_rows = Int(foldl((x, y) -> x * length(y), uniqueVals, init=big(1)))
+    if iszero(target_rows)
+        @assert iszero(nrow(df))
+        return allcols ? select(df, colind, :) : select(df, colind)
+    end
+
+    # construct expanded columns
+    out_df = DataFrame()
+    inner = 1
+    @assert length(uniqueVals) == length(colind)
+    for (val, cind) in zip(uniqueVals, colind)
+        len = length(val)
+        target_col = Tables.allocatecolumn(eltype(df[!, cind]), len)
+        target_col .= val
+        last_inner = inner
+        inner *= len
+        outer, remv = divrem(target_rows, inner)
+        @assert iszero(remv)
+        out_df[!, _names(df)[cind]] = repeat(target_col, inner=last_inner, outer=outer)
+    end
+    @assert inner == target_rows
+
+    # optionally join the remaining columns
+    if allcols
+        idx_ind = 0
+        while columnindex(df, string("source", idx_ind)) != 0
+            idx_ind += 1
+        end
```

Comment:
These uncovered lines appear to be real?

Reply:
this is covered.